lte/gateway/c/oai/tasks/ha/ha_service_handler.cpp

```diff
@@ -0,0 +1,179 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include <iostream>
+#include <string.h>
+#include <sys/types.h>
+
+extern "C" {
+#include "common_types.h"
+#include "ha_defs.h"
+#include "ha_messages_types.h"
+#include "intertask_interface.h"
+#include "intertask_interface_types.h"
+#include "itti_types.h"
+#include "log.h"
+#include "s1ap_state.h"
+#include "S1ap_CauseRadioNetwork.h"
+}
+
+#include "HaClient.h"
+#include "mme_app_state_manager.h"
+#include "s1ap_state_manager.h"
+
+static bool process_ue_context(
+    const hash_key_t keyP, void* const elementP, void* parameterP,
+    void** resultP);
+
+bool sync_up_with_orc8r(void) {
+  magma::HaClient::get_eNB_offload_state(
+      [](grpc::Status status,
+         magma::lte::GetEnodebOffloadStateResponse response) {
+        if (status.ok()) {
+          OAILOG_INFO(
+              LOG_UTIL, "Received eNodeB connection state with the primary.");
+          // iterate over the eNodeB connection states
+          ha_agw_offload_req_t offload_req = {0};
+          for (auto const& item : response.enodeb_offload_states()) {
+            if (item.second ==
+                magma::lte::GetEnodebOffloadStateResponse::PRIMARY_CONNECTED) {
+              offload_req.eNB_id = item.first;
+              // Offload any UE to check if it can be camped on the primary.
+              // The effect will be observed in the next sync up with the cloud.
+              offload_req.enb_offload_type = ANY;
+              handle_agw_offload_req(&offload_req);
+            } else if (
+                item.second == magma::lte::GetEnodebOffloadStateResponse::
+                                   PRIMARY_CONNECTED_AND_SERVING_UES) {
+              offload_req.eNB_id = item.first;
+              // Primary looks healthy as UEs are camped on it, offload the rest
+              // of UEs.
+              offload_req.enb_offload_type = ALL;
+              handle_agw_offload_req(&offload_req);
+            }
+          }
+        } else {
+          OAILOG_ERROR(
+              LOG_UTIL, "GRPC Failure Message: %s Status Error Code: %d",
+              status.error_message().c_str(), status.error_code());
+        }
+      });
+  return true;
+}
+
+typedef struct callback_data_s {
+  s1ap_state_t* s1ap_state;
+  ha_agw_offload_req_t* request;
+} callback_data_t;
+
+void handle_agw_offload_req(ha_agw_offload_req_t* offload_req) {
+  hash_table_ts_t* state_imsi_ht =
+      magma::lte::MmeNasStateManager::getInstance().get_ue_state_ht();
+  callback_data_t callback_data;
+  callback_data.s1ap_state =
+      magma::lte::S1apStateManager::getInstance().get_state(false);
+  callback_data.request = offload_req;
+  hashtable_ts_apply_callback_on_elements(
+      state_imsi_ht, process_ue_context, (void*) &callback_data, NULL);
+}
+
+bool process_ue_context(
```

Comment:
Can we use a function name which is more informative, e.g.
```suggestion
bool trigger_agw_offload_for_ue(
```

Reply:
done.